counsel.el

```diff
@@ -3887,6 +3887,56 @@ Note: Duplicate elements of `kill-ring' are always deleted."
  '(("d" counsel-yank-pop-action-remove "delete")
    ("r" counsel-yank-pop-action-rotate "rotate")))
 
+;;** `counsel-register'
+(defvar counsel-register-actions
+  '(("\\`a buffer position" . jump-to-register)
+    ("\\`text" . insert-register)
+    ("\\`a rectangle" . insert-register)
+    ("\\`a window configuration" . jump-to-register)
+    ("\\`a frame configuration" . jump-to-register)
+    ("\\`[-+]?[0-9]+\\(?:\\.[0-9]\\)?\\'" . insert-register)
+    ("\\`the file" . jump-to-register)
+    ("\\`a keyboard macro" . jump-to-register)
+    ("\\`a file-query" . jump-to-register))
+  "Alist of (REGEXP . FUNCTION) pairs for `counsel-register'.
+Selecting a register whose description matches REGEXP specifies
+FUNCTION as the action to take on the register.")
+
+(defvar counsel-register-history nil
+  "History for `counsel-register'.")
+
+(defun counsel-register-action (register)
+  "Default action for `counsel-register'.
+
+Call a function on REGISTER.  The function is determined by
+matching the register's value description against a regexp in
+`counsel-register-actions'."
+  (setq register (string-to-char register))
+  (let* ((desc (with-output-to-string
+                 (register-val-describe (get-register register) nil)))
+         (action (cl-assoc-if (lambda (re)
+                                (string-match-p re desc))
+                              counsel-register-actions)))
+    (if (cdr action)
+        (funcall (cdr action) register)
+      (error "No action was found for register %c" register))))
+
+;;;###autoload
+(defun counsel-register ()
+  "Interactively choose a register."
+  (interactive)
+  (ivy-read "Register: "
+            (mapcar (lambda (reg)
+                      (let* ((s (funcall register-preview-function reg))
+                             (i (string-match-p "\\(?:[ \t\n\r]+\\)\\'" s)))
```

Comment:
No need for the shy group: `"[ \t\n\r]+\\'"`

I think a comment like the following would also be nice:

```diff
diff --git a/counsel.el b/counsel.el
index bd1a4aa..b69b52f 100644
--- a/counsel.el
+++ b/counsel.el
@@ -3927,6 +3927,7 @@ counsel-register
   (interactive)
   (ivy-read "Register: "
             (mapcar (lambda (reg)
+                      ;; `string-trim' was only added in Emacs 24.4.
                       (let* ((s (funcall register-preview-function reg))
                              (i (string-match-p "\\(?:[ \t\n\r]+\\)\\'" s)))
                         (if i (substring s 0 i) s)))
```

Alternatively you could define a compatibility shim, but it's probably overkill:

```el
(defalias 'counsel--trim-right
  (if (and (require 'subr-x nil t)
           (fboundp 'string-trim-right))
      #'string-trim-right
    (lambda (string)
      (let ((i (string-match-p "[ \t\n\r]+\\'" string)))
        (if i (substring string 0 i) string))))
  "Trim STRING of trailing whitespace.

\(fn STRING)")
```